### ISSUE TITLE
MQTT TLS with Hostname don't work in 9.5 but works on 9.4 (not core related)

### ISSUE THREAD

**Issue by bwmtx** [arendst/Tasmota#12988]
DO NOT DELETE ANY TEXT from this template! Otherwise the issue will be auto-closed.
-->

### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

I desire to upgrade from 9.3 to 9.5 so to take advantage of the new user_rule[x] and user_backlog features.

My problem is the TLS connection to my server with LetsEncrypt cert fails.  I get multiple consecutive connections the socket errors.

`New connection from [ip address] on port [port]`
`Scocket error on client <unknown>, disconnecting`

With tls enabled the esp nodemcu keeps rebooting (boot loop??).  I checked  my router logs... the device constantly requests dhcp address.

If I flash 9.5 with local mqtt setting (tls disabled) everything works as expected

I also tried to brand new esp modules.  Same behavior.

Version 9.3 connects just fine.  

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): NodeMCU ESP8266MOD
- [x] Tasmota binary firmware version number used: 9.5
  - [ ] Pre-compiled
  - [x] Self-compiled   
     Built using vs code platformio
     `MQTT_TLS_ENABLE    true
      #define USE_MQTT_TLS
        #define USE_MQTT_TLS_CA_CERT
     `
- [x] Flashing tools used: esptool.py
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
Configuration output here:  Mqtt and web interface aren't accessible.  I get 2-3 ping responses then 9-10 host unavailable.
```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
Rules output here:  I have rules defined in the firmware, but have commented them out until the connection issue is resolved.
```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:  Mqtt and web interface aren't accessible.
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here: Mqtt and web interface aren't accessible.
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Enable tls pointed to server with letsencrypt certs.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Tls connection to cloud server working

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by bwmtx** [comment]
@ascillato2 

How long do you think it'll take to get this resolved?  So I don't anxiously keep checking :). Thanks

**Comment by ascillato** [comment]
> How long do you think it'll take to get this resolved? So I don't anxiously keep checking :). Thanks

hahahah. :) I really don't know. sorry. Anyway, if you are subscribed to this issue, every time someone writes here, GitHub should send an email to you.

**Comment by bwmtx** [comment]
I compiled version 9.4 with same settings... it works as expected.

**Comment by bwmtx** [comment]
So I did some more digging.

Pings to the device stabilized once I turned off mqtt (setoption3 0).  Gave it a few mins the turned back on mqtt. See below...

tasmota console --> MQT: Connect failed to <my_domain>:8883, rc -4. Retry in 10 sec

/var/log/mosquitto/mosquitto.log --> ssl3_get_record:wrong version number

Accordning to https://tasmota.github.io/docs/TLS/ error code 4 is -->  BR_ERR_BAD_VERSION, which corresponds to the mosquitto.log message.

**Comment by arendst** [comment]
> Pings to the device stabilized once I turned off mqtt (setoption3 0). Gave it a few mins the turned back on mqtt.

As designed. As long as MQTT tries to connect to the server all other communication is blocked until MQTT tiemout is expired. So you won't see any response of your pings during this time.

> tasmota console --> MQT: Connect failed to <my_domain>:8883, rc -4. Retry in 10 sec

-4 : MQTT_CONNECTION_TIMEOUT      - the server didn't respond within the keepalive time

> error code 4 is --> BR_ERR_BAD_VERSION

4 : 260 : 516 : BR_ERR_BAD_VERSION         - incoming record version does not match the expected version

**Comment by bwmtx** [comment]
Got it.  So what changed in Ver 1.5 that would cause this behavior?  It's strange because Ver. 1.4 connects with the same settings.

It's almost like the issue has to do with the clientid or user/pass because...

a successful connection using Ver 1.4 --> 
```
New connection from <public_id> on port 8883.
New client connected from <public_ip> as DVES_E30DC1 (p2, c1, k30, u'user').
```

a failed connection using Ver. 1.5 -->
```
New connection from <public_ip> on port 8883.
New connection from <public_ip> on port 8883.
New connection from <public_ip> on port 8883.
Client <unknown> has exceeded timeout, disconnecting.
Client <unknown> has exceeded timeout, disconnecting.
Client <unknown> has exceeded timeout, disconnecting.
```

Below is my mosquitto.conf
```
pid_file /var/run/mosquitto.pid

persistence true
persistence_location /var/lib/mosquitto/
##autosave_interval 900
check_retain_source true

log_dest file /var/log/mosquitto/mosquitto.log
#log_type all
allow_anonymous false
password_file /etc/mosquitto/passwd
acl_file /etc/mosquitto/acl

allow_zero_length_clientid false

include_dir /etc/mosquitto/conf.d

listener 1883 localhost

listener 8883
certfile /etc/letsencrypt/live/mydomain.com/cert.pem
cafile /etc/letsencrypt/live/mydomain.com/chain.pem
keyfile /etc/letsencrypt/live/mydomain.com/privkey.pem
```

**Comment by mrkeuz** [comment]
Faced with same issue when compile 9.5.0 with custom hostname. 
If I remember correctly also use flag "MQTT TLS" also not working properly from UI.
It cannot change from UI. And need setup via command.

My WORKAROUND is just:
Set hostname other settings via commands 
Without comments and new lines, and one space after semicolons: 
```python
Backlog 
MqttPort 8883;
MqttUser MQTT_USER;
MqttPassword MQTT_PASSWORD;
MqttHost MQTT_HOST;
# Enable MQTT TLS;
SetOption103 MQTT_USE_TLS;
# Reset device
Restart 1
```

Just in case. If it's stopper for someone. 
Work fine with third level domain and for plain IP address.

**Comment by bwmtx** [comment]
@mrkeuz 
Thanks for the workaround solution!  I may give it a go.

I've downgraded to 9.4 for the time being.  Have been tracking the dev branch for updates.

**Comment by Arakula** [comment]
I ran into the same problem.

After the downgrade to 9.4.0, TLS worked ... kind of. It didn't lead to permanent "FRC: Some settings have been reset" any more; instead, I got a nice "MQT: TLS connection error: 23". Documentation says: "BR_ERR_LIMIT_EXCEEDED: the server's public key is too large. Tasmota TLS is limited to 2048 RSA keys". Oooo-KAY ... I uncommented the -DUSE_4K_RSA setting in platformio.ini, section [esp82xx_defaults], then cleaned the cache, rebuilt and ... HEY! It WORKS!
Good.

I then applied the same change to platformio.ini in 9.5.0, and all of a sudden, this one worked, too. Just to be sure that it wasn't just caused by the preceding 9.4.0 setup, I did a configuration reset, set everything up again, and ... YES!

So: the additional -DUSE_4K-RSA was necessary in my case. Maybe it does it for you, too.
